<div style="border:solid green 2px; padding: 20px"> 
    
**Комментарий ревьюера**

Привет! Меня зовут Сысоева Вера.
    
Далее в файле мои комментарии ты сможешь найти в ячейках, аналогичных данной ( если рамки комментария зелёные и отмечены **#ок** - всё сделано правильно; жёлтые и отмечены **#можно лучше** - есть замечания, но не критично; красные с пометкой **#ошибка** - нужно переделать). Не удаляй эти комментарии и постарайся учесть их в ходе выполнения проекта.
    
Я буду поддерживать версионность комментариев, и при следующих итерациях комментарии будут выделены, как "Комментраии ревьюера - 2 (3 и.т.д.)"
    
Если у тебя есть замечания, возражения и вопросы, ты можешь писать свои комментарии, по возможности выделяй их так же цветом (например, синим), чтоб их можно было легко отделить от текста проекта. 
    
Кстати, предлагаю сразу общаться на "ты".
    
Увидев у тебя ошибку, в первый раз я лишь укажу на ее наличие и дам тебе возможность самостоятельно найти и исправить ее. На реальной работе твой начальник будет поступать так же, а я пытаюсь подготовить тебя именно к работе аналитиком. Но если ты пока не справишься с такой задачей - при следующей проверке я дам более точную подсказку.
    
</div>

<div class="alert alert-info">
<b> Комментарий от студента - 2 </b>
    
Привет! Все замечания поправил. Появилось несколько небольших вопросов, буду рад получить ответы на них!)
</div>

# Исследование надежности заемщиков.


**Описание проекта**

Заказчик — кредитный отдел банка. Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Входные данные от банка — статистика о платёжеспособности клиентов.
Результаты исследования будут учтены при построении модели кредитного скоринга — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.

**Цель исследования** — проверка следующих гипотез:

- Есть ли зависимость между количеством детей и возвратом кредита в срок?
- Есть ли зависимость между семейным положением и возвратом кредита в срок?
- Есть ли зависимость между уровнем дохода и возвратом кредита в срок?
- Как разные цели кредита влияют на его возврат в срок?

# Шаг 1. Обзор данных

<div class="alert alert-info">
<b> Комментарий от студента </b>
    
Вопрос: в прошлом проекте упоминалось, что вывод таблиц с помощью print() лучше заменить на display() для более корректного отображения данных в JN. Есть ли необходимость загонять, например, df.head(10) внутрь display(), если и без этого она отображается корректно.
</div>

<div style="border:solid green 2px; padding: 20px"> 
    
**Комментарий ревьюера**

display используется в jupiter автоматически для последней строки кода в ячейке. Дополнительно его писать не нужно. Но, если мы хотим вывести не последнюю строку (часто это в цикле используется), то для вывода таблиц стоит использовать display, а не print
</div>

In [140]:
# импорт библиотеки pandas и math

import pandas as pd
import math

<div style="border:solid green 2px; padding: 20px"> 
    
**Комментарий ревьюера**

Правильно, что ты подгружаешь библиотеки в отдельной первой ячейке
</div>

In [141]:
# чтение данных из файла .csv и сохранение в df

df = pd.read_csv('/datasets/data.csv')

<div style="border:solid green 2px; padding: 20px"> 
    
**Комментарий ревьюера**

**#ок**

Отлично, все верно подгружено. 
    
Также при работе с проектами тебе может понадобится работать на своей машине локально. Но в итоге сдать его можно только с сервера. Чтобы данные подгружались и локально и глобально можно использовать конструкцию try-except или if-else

Также в таких случаях стоит применять библиотеку ``os``:

    import os

    pth1 = '/folder_1/data.csv' # Локальный путь
    pth2 = '/folder_2/data.csv' # Серверный путь
    
    if os.path.exists(pth1):
        data = pd.read_csv(pth1)
    elif os.path.exists(pth2):
        data = pd.read_csv(pth2)
    else:
        print('Something is wrong)  
    
Т.е. программа пытается подгрузить файл локально, и если не удается, то подгружает с сервера, если и это не удается, то пишет об ошибке
    
</div>

In [142]:
# получение первых 10 строк таблицы df

df.head(10)

children  days_employed  dob_years education  education_id  \
0         1   -8437.673028         42    высшее             0   
1         1   -4024.803754         36   среднее             1   
2         0   -5623.422610         33   Среднее             1   
3         3   -4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   
5         0    -926.185831         27    высшее             0   
6         0   -2879.202052         43    высшее             0   
7         0    -152.779569         50   СРЕДНЕЕ             1   
8         2   -6929.865299         35    ВЫСШЕЕ             0   
9         0   -2188.756445         41   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   
5  гражданский брак                 1      M   компаньон     0  255763.565419   
6   женат / замужем                 0      F   компаньон     0  240525.971920   
7   женат / замужем                 0      M   сотрудник     0  135823.934197   
8  гражданский брак                 1      F   сотрудник     0   95856.832424   
9   женат / замужем                 0      M   сотрудник     0  144425.938277   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу  
5               покупка жилья  
6           операции с жильем  
7                 образование  
8       на проведение свадьбы  
9     покупка жилья для семьи

In [143]:
# вывод общей информации по таблице df

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       19351 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        19351 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


Согласно полученной информации, в таблице 12 столбцов. Тип данных в столбцам int, float и object.

Согласно полученной документации:
* `children` — количество детей в семье
* `days_employed` — общий трудовой стаж в днях
* `dob_years` — возраст клиента в годах
* `education` — уровень образования клиента
* `education_id` — идентификатор уровня образования
* `family_status` — семейное положение
* `family_status_id` — идентификатор семейного положения
* `gender` — пол клиента
* `income_type` — тип занятости
* `debt` — имел ли задолженность по возврату кредитов
* `total_income` — ежемесячный доход
* `purpose` — цель получения кредита

В названиях колонок отсутствует нарушения стиля.

Количество значений в столбцах `days_employed` и `total_income` отличается от остальных столбцов. Следовательно, в данных есть пропущенные значения.

Также в данных присутствуют аномалии, такие как отрицательный трудовой стаж клиента.

<div style="border:solid green 2px; padding: 20px"> 
    
**Комментарий ревьюера**

**#ок**

Данные верно загружены, правильно просмотрена общая информация. Верные методы использованы, молодец

</div>

**Выводы**

В каждой строке таблицы — данные о клиентах банка. Колонки рассказывают о клиенте: количество детей, трудовой стаж, семейное положение, уровень дохода и пр.

Предварительно можно утверждать, что данных достаточно для проверки обозначенных гипотез. Однако в данных встречаются пропуски. Также необходимо проверить полученные данные на наличие аномалий, явных и неявных дубликатов.

Чтобы двигаться дальше необходимо устранить проблемы в данных.

# Шаг 2.1 Заполнение пропусков

In [144]:
# Посчитаем число пропущенных значений в таблице

df.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

In [145]:
# Находим все строки с пропусками в столбце days_employed и выводим первые десять

df[df['days_employed'].isna()].head(10)

children  days_employed  dob_years education  education_id  \
12         0            NaN         65   среднее             1   
26         0            NaN         41   среднее             1   
29         0            NaN         63   среднее             1   
41         0            NaN         50   среднее             1   
55         0            NaN         54   среднее             1   
65         0            NaN         21   среднее             1   
67         0            NaN         52    высшее             0   
72         1            NaN         32    высшее             0   
82         2            NaN         50    высшее             0   
83         0            NaN         52   среднее             1   

            family_status  family_status_id gender  income_type  debt  \
12       гражданский брак                 1      M    пенсионер     0   
26        женат / замужем                 0      M  госслужащий     0   
29  Не женат / не замужем                 4      F    пенсионер     0   
41        женат / замужем                 0      F  госслужащий     0   
55       гражданский брак                 1      F    пенсионер     1   
65  Не женат / не замужем                 4      M    компаньон     0   
67        женат / замужем                 0      F    пенсионер     0   
72        женат / замужем                 0      M  госслужащий     0   
82        женат / замужем                 0      F    сотрудник     0   
83        женат / замужем                 0      M    сотрудник     0   

    total_income                                purpose  
12           NaN                        сыграть свадьбу  
26           NaN                            образование  
29           NaN       строительство жилой недвижимости  
41           NaN       сделка с подержанным автомобилем  
55           NaN                        сыграть свадьбу  
65           NaN  операции с коммерческой недвижимостью  
67           NaN                покупка жилья для семьи  
72           NaN  операции с коммерческой недвижимостью  
82           NaN                                  жилье  
83           NaN                                  жилье

Можно заметить, что в столбцах `days_employed` и `total_income` присутствуют значения NaN. 

In [146]:
# Найдем долю пропущенных значений в столбцах days_employed и total_income

days_employed_miss = len(df[df['days_employed'].isna()])
total_income_miss = len(df[df['total_income'].isna()])
total_row_count = df.shape[0]

print(f'Доля пропущенных значений в столбце days_employed составляет: {round(days_employed_miss / total_row_count * 100, 1)}%')
print(f'Доля пропущенных значений в столбце total_income составляет: {round(total_income_miss / total_row_count * 100, 1)}%')

Доля пропущенных значений в столбце days_employed составляет: 10.1%
Доля пропущенных значений в столбце total_income составляет: 10.1%


Доля пропущенных значений составляет приблизительно 10%. Данные в столбцах `days_employed` и `total_income` являются количественными, следовательно, в сложившейся ситуации для недопущения потери ценных данных решено заменить пропущенные значения на медианное значение по каждому из стоблцов. 

Медианное значение выбирается с целью более качественной характеристики данных, т.к. среднее значение не всегда корректно характеризует данные, когда некоторые значения исходного пула сильно выделяются среди большинства.

Причиной появления пропусков данных может быть связана с различной формой сбора данных, либо ошибке при их выгрузке. В поддержку данной гипотезы выступает тот факт, что число пропущенных строк в двух столбцах одинаковы.

In [147]:
# Заменим пропущенные значения в столбцах days_employed и total_income

days_avg = df['days_employed'].median()
income_avg = df['total_income'].median()

print('Медианное значение для трудового стажа:', days_avg)
print('Медианное значение для заработка:', income_avg)

df['days_employed'] = df['days_employed'].fillna(value=days_avg)
df['total_income'] = df['total_income'].fillna(value=income_avg)

Медианное значение для трудового стажа: -1203.369528770489
Медианное значение для заработка: 145017.93753253992


<div style="border:solid green 2px; padding: 20px"> 
    
**Комментарий ревьюера**

**#ок**

Заполнение медианой - правильное решение. В данных есть выбросы, а медиана наименее чувствительна к ним

</div>

In [148]:
# Проверим результат замены NaN пропусков

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       21525 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        21525 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


Как мы видим, пропуски значений заменились нашими медианными значениями. Следовательно, мы готовы к дальнейшей предобработке данных.

# Шаг 2.2 Проверка данных на аномалии и исправления.

Обратим внимание, что медианное значение по столбцу `days_employed` - отрицательное. Это аномалия. Большинство значений в столбце `days_employed` представлено отрицательными значениями или значениями, выпадающими из возможного пула (например, в строчке 4: 340266.072047). Согласно данному значению общий трудовой стаж человека составил 932 года. Возможно, при заполнении данных трудовой стаж был указан в часах, а не днях, как заявлено в сопроводительной документации. Либо была допущена ошибка при заполнении данных. Также большая часть трудового стажа указана с отрицательным знаком. Вероятнее всего, это технологическая ошибка и стоит доложить о ней разработчику, который прислал исходный файл для уточнения корректности данных.

Однако, согласно требованиям заказчика, работ со столбцом `days_employed` не предвидится и необходимости в разборе аномалий по данном столбцу нету. Поэтому в дальнейших исследованиях будем иметь ввиду некорректность столбца `days_employed`, но тратить время на его обработку не будем.

In [149]:
# Выведем первые 20 строчек таблицы для визуальной проверки замены и контроля аномальных значений в столбцах

df.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0   -1203.369529         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12  145017.937533                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

Сразу обратим внимание, что в таблице присутствует большое количество дубликатов:
- разный регистр в столбце `education`;
- дубликаты категорий в столбце `purpose`;
- некорректный тип данных в столбце `total_income`.

Разберемся с этими проблемами далее.

<div style="border:solid gold 2px; padding: 20px"> 
    
**Комментарий ревьюера**

**#можно лучше**

Для поиска ошибок стоило прямо в этом разделе использовать `value_counts` для категориальных переменных и `describe` для получения информации о количественных переменных.

Ты делаешь это ниже вполне правильно, но стоило именно здесь искать так ошибки. Сейчас кажется, что ты хочешь найти ошибки, посмотрев первые 20 строк, а это принципиально не верно. Никакие 20, 50 или даже 100 первых строк не могут служить проверкой наличия или исправления ошибок. Нужно использовать другие методы: смотреть подвыборки, считать количества. В общем надо точно показывать сколько у нас строк был с такой-то ошибкой и сколько стало (должно быть соответственно 0)
</div>

<div class="alert alert-info">
<b> Комментарий от студента - 2</b>
    
Я хотел визуально оценить то, как выглядят данные в таблице. Но теперь осознаю, что это может случайно исказить реальную картину и так лучше не делать. Блок удалять не буду, но ошибку понял.
</div>

<div style="border:solid green 2px; padding: 20px"> 
    
**Комментарий ревьюера - 2**

👍
    
</div>

# Шаг 2.3. Изменение типов данных.

In [150]:
# Переведем все значения столбца `total_income` в тип int и убедимся в корретности замены

df['total_income'] = df['total_income'].astype('int')
df['days_employed'] = df['days_employed'].astype('int')
df['debt'] = df['debt'].astype('bool')
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       21525 non-null int64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null bool
total_income        21525 non-null int64
purpose             21525 non-null object
dtypes: bool(1), int64(6), object(5)
memory usage: 1.8+ MB


<div style="border:solid red 2px; padding: 20px"> 
    
**Комментарий ревьюера**

**#ошибка**

Также нужно стаж (days_employed) сделать Int и debt сделать bool, тогда типы данных будут соответствовать содержимому и станут оптимальны
</div>

<div class="alert alert-info">
<b> Комментарий от студента - 2</b>
    
Проглядел эти моменты. Исправил.
</div>

<div style="border:solid green 2px; padding: 20px"> 
    
**Комментарий ревьюера - 2**

+
    
</div>

# Шаг 2.4. Удаление дубликатов.

In [151]:
# Посчитаем количество дубликатов в таблице

#print('Дубликатов в таблице:', df.duplicated().sum())

In [152]:
# Удаляем все дубликаты в таблице

#df = df.drop_duplicates().reset_index(drop=True)

In [153]:
# Проверяем число дубликатов после удаления

#print('Дубликатов в таблице после обработки:', df.duplicated().sum())

<div style="border:solid red 2px; padding: 20px"> 
    
**Комментарий ревьюера**

**#ошибка**

Технически ты верно поработал с дубликатами. Только явные дубликаты нужно считать и удалять уже после всех преобразований, т.к. исправление ошибок может привести к увеличению числа дубликатов.
    
Перенеси эту часть в самый конец данного раздела и посмотри сколько тогда будет дублей
</div>

<div class="alert alert-info">
<b> Комментарий от студента - 2</b>
    
Да, ты оказалась права. Количество явных дубликатов после процедур предобработки данных стало на несколько десятков больше. Закомменил предыдущие шаги, чтобы не нарушать ход проверки. Содержимое перенес в конец раздела. Спасибо, важное замечание!
</div>

Вспоминаем, что на прошлом шаге мы заметили, что в таблице присутствуют неявные дуликаты. Пройдемся по всем нужным нам столбцам для поиска неявных дубликатов и заменим их значения.

In [154]:
# Выводим все уникальные значения по столбцу `dob_years`

df_sorted = df.sort_values(by='dob_years')
df_sorted['dob_years'].unique()

array([ 0, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34,
       35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51,
       52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68,
       69, 70, 71, 72, 73, 74, 75])

Обращаем внимание на первое значение в списке значений. Возвраст заемщика - 0 лет. Найдем все строки таблицы, где возраст заемщика 0 лет.

In [155]:
display(df.loc[df.loc[:,'dob_years'] == 0])

children  days_employed  dob_years education  education_id  \
99            0         346541          0   Среднее             1   
149           0          -2664          0   среднее             1   
270           3          -1872          0   среднее             1   
578           0         397856          0   среднее             1   
1040          0          -1158          0    высшее             0   
...         ...            ...        ...       ...           ...   
19829         0          -1203          0   среднее             1   
20462         0         338734          0   среднее             1   
20577         0         331741          0   среднее             1   
21179         2           -108          0    высшее             0   
21313         0          -1268          0   среднее             1   

               family_status  family_status_id gender income_type   debt  \
99           женат / замужем                 0      F   пенсионер  False   
149                в разводе                 3      F   сотрудник  False   
270          женат / замужем                 0      F   сотрудник  False   
578          женат / замужем                 0      F   пенсионер  False   
1040               в разводе                 3      F   компаньон  False   
...                      ...               ...    ...         ...    ...   
19829        женат / замужем                 0      F   сотрудник  False   
20462        женат / замужем                 0      F   пенсионер  False   
20577  Не женат / не замужем                 4      F   пенсионер  False   
21179        женат / замужем                 0      M   компаньон  False   
21313  Не женат / не замужем                 4      M   сотрудник  False   

       total_income                                 purpose  
99            71291                              автомобиль  
149           70176                       операции с жильем  
270          102166                            ремонт жилью  
578           97620  строительство собственной недвижимости  
1040         303994                         свой автомобиль  
...             ...                                     ...  
19829        145017                                   жилье  
20462        259193                    покупка своего жилья  
20577        129788                            недвижимость  
21179        240702        строительство жилой недвижимости  
21313        152419                       операции с жильем  

[101 rows x 12 columns]

Всего строк со значением возраста заёмщика 0 лет - 101 штука. Данные этих строк могут быть полезны для наших исследований, т.к. в исследованиях не учитывается возраст заёмщика. Найдем среднее значение возраста заёмщиков и заменим нулевые значения на среднее.

In [156]:
# Находим среднее значение возраста заемщика и заменяем значения 0 на среднее значение

dob_years_avg = int(df['dob_years'].mean())
print('Среднее значение для возраста заемщика:', dob_years_avg)

df['dob_years'] = df['dob_years'].replace(0, dob_years_avg)

Среднее значение для возраста заемщика: 43


<div style="border:solid green 2px; padding: 20px"> 
    
**Комментарий ревьюера**

**#ок**

Отлично, с 0-ми верно разобрались
</div>

In [157]:
# Проверим результат замены 0 на средний возраст

df_sorted = df.sort_values(by='dob_years')
df_sorted['dob_years'].unique()

array([19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35,
       36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52,
       53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69,
       70, 71, 72, 73, 74, 75])

Столбец `dob_years` приведен в соответствие.

In [158]:
# Выводим все уникальные значения по столбцу `children`

df_sorted = df.sort_values(by='children')
df_sorted['children'].unique()

array([-1,  0,  1,  2,  3,  4,  5, 20])

Обращаем внимание на первое значение в списке. Количество детей — "-1". Найдем все строки, где количество детей -1.

In [159]:
# Найдем все строки, где количество детей -1.

display(df.loc[df.loc[:,'children'] == -1])

children  days_employed  dob_years            education  education_id  \
291          -1          -4417         46              среднее             1   
705          -1           -902         50              среднее             1   
742          -1          -3174         57              среднее             1   
800          -1         349987         54              среднее             1   
941          -1          -1203         57              Среднее             1   
1363         -1          -1195         55              СРЕДНЕЕ             1   
1929         -1          -1461         38              среднее             1   
2073         -1          -2539         42              среднее             1   
3814         -1          -3045         26              Среднее             1   
4201         -1           -901         41              среднее             1   
4402         -1         398001         64              СРЕДНЕЕ             1   
4542         -1          -1811         32              среднее             1   
5272         -1          -1143         46              среднее             1   
6013         -1          -1361         46               высшее             0   
6385         -1         370215         48              среднее             1   
7190         -1          -5928         34              СРЕДНЕЕ             1   
7278         -1           -526         51              среднее             1   
7615         -1          -1203         35              среднее             1   
7685         -1          -3237         53              среднее             1   
8229         -1          -1803         54               высшее             0   
8251         -1          -3113         31              среднее             1   
8561         -1           -195         31              среднее             1   
9565         -1          -2896         34               высшее             0   
9574         -1          -2710         28               высшее             0   
10177        -1          -1743         37              среднее             1   
10365        -1         345774         63               ВЫСШЕЕ             0   
10918        -1         340499         50              среднее             1   
11109        -1          -3438         59              среднее             1   
11268        -1          -1048         30               высшее             0   
11617        -1           -370         27              среднее             1   
12304        -1          -8493         61  неоконченное высшее             2   
13786        -1          -1203         42              среднее             1   
14357        -1           -268         23              среднее             1   
15147        -1          -9851         44              среднее             1   
15435        -1          -3614         40               высшее             0   
16129        -1           -457         33              среднее             1   
16265        -1          -2802         40               высшее             0   
17095        -1          -2809         34              среднее             1   
17429        -1           -895         37              среднее             1   
17657        -1          -4571         41              среднее             1   
18219        -1          -3575         33              среднее             1   
19100        -1           -617         38              среднее             1   
19366        -1           -372         43              СРЕДНЕЕ             1   
19417        -1         350340         28              среднее             1   
20393        -1         355157         69              среднее             1   
20717        -1           -661         32               высшее             0   
21140        -1          -1422         44              среднее             1   

               family_status  family_status_id gender  income_type   debt  \
291         гражданский брак                 1      F    сотрудник  False   
705         

In [160]:
# Посчитаем количество строк со значением "-1" в столбце с количеством детей

display(df.loc[df.loc[:,'children'] == -1].count())

children            47
days_employed       47
dob_years           47
education           47
education_id        47
family_status       47
family_status_id    47
gender              47
income_type         47
debt                47
total_income        47
purpose             47
dtype: int64

Мы видим, что число строк с некорретным значением всего 47 штук. Ввиду отсутствия возможности в рамках данной работы уточнить у разработчика расшифровку значения "-1" в столбце с количеством детей принимает решение о удалении строк с некорректным значением для сохранения целостности итоговых результатов.

<div class="alert alert-info">
<b> Комментарий от студента </b>
    
Этот момент мне бы хотелось, чтобы Вы разъяснили правильно ли я сделал, что принял решение об удалении этих данных с целью не внесения погрешности в общий пул результатов. Спасибо!
</div>

<div style="border:solid green 2px; padding: 20px"> 
    
**Комментарий ревьюера**

**#ок**

Удалить -1 в данном случае возможное решение, их достаточно мало (в таких случаях хорошо показать процент строк).

</div>

In [161]:
# Удалим из df строки, в которых количество детей равно -1

df = df.loc[df['children'] != -1].reset_index(drop=True)
display(df)

children  days_employed  dob_years education  education_id  \
0             1          -8437         42    высшее             0   
1             1          -4024         36   среднее             1   
2             0          -5623         33   Среднее             1   
3             3          -4124         32   среднее             1   
4             0         340266         53   среднее             1   
...         ...            ...        ...       ...           ...   
21473         1          -4529         43   среднее             1   
21474         0         343937         67   среднее             1   
21475         1          -2113         38   среднее             1   
21476         3          -3112         38   среднее             1   
21477         2          -1984         40   среднее             1   

          family_status  family_status_id gender income_type   debt  \
0       женат / замужем                 0      F   сотрудник  False   
1       женат / замужем                 0      F   сотрудник  False   
2       женат / замужем                 0      M   сотрудник  False   
3       женат / замужем                 0      M   сотрудник  False   
4      гражданский брак                 1      F   пенсионер  False   
...                 ...               ...    ...         ...    ...   
21473  гражданский брак                 1      F   компаньон  False   
21474   женат / замужем                 0      F   пенсионер  False   
21475  гражданский брак                 1      M   сотрудник   True   
21476   женат / замужем                 0      M   сотрудник   True   
21477   женат / замужем                 0      F   сотрудник  False   

       total_income                       purpose  
0            253875                 покупка жилья  
1            112080       приобретение автомобиля  
2            145885                 покупка жилья  
3            267628    дополнительное образование  
4            158616               сыграть свадьбу  
...             ...                           ...  
21473        224791             операции с жильем  
21474        155999          сделка с автомобилем  
21475         89672                  недвижимость  
21476        244093  на покупку своего автомобиля  
21477         82047         на покупку автомобиля  

[21478 rows x 12 columns]

In [162]:
# Убедимся в отсутствии строк со значением количества детей "-1"

df_sorted = df.sort_values(by='children')
df_sorted['children'].unique()

array([ 0,  1,  2,  3,  4,  5, 20])

Столбец `childern` приведен в соответствие.

In [163]:
# Выводим все уникальные значения по столбцу `education`

print(df['education'].value_counts()) 

среднее                13719
высшее                  4710
СРЕДНЕЕ                  768
Среднее                  709
неоконченное высшее      667
ВЫСШЕЕ                   273
Высшее                   268
начальное                250
Неоконченное высшее       47
НЕОКОНЧЕННОЕ ВЫСШЕЕ       29
НАЧАЛЬНОЕ                 17
Начальное                 15
ученая степень             4
УЧЕНАЯ СТЕПЕНЬ             1
Ученая степень             1
Name: education, dtype: int64


In [164]:
# Устраним дубликаты из-за разниц регистра

df['education'] = df['education'].str.lower()
display(df['education'].value_counts()) 

среднее                15196
высшее                  5251
неоконченное высшее      743
начальное                282
ученая степень             6
Name: education, dtype: int64

В данном случае с помощью .str.lower() мы однострочно устранили все неявные дубликаты, связанные с разным написанием регистра. Причина его возникновения, скорее всего, связана с отсутствие дополнительной обработки данных после форм, заполняемых заемщиком. Об этом стоит сообщить команде разработчиков для устранения этой проблемы.

In [165]:
# Выводим все уникальные значения по столбцу `education_id`

print(df['education_id'].value_counts()) 

1    15196
0     5251
2      743
3      282
4        6
Name: education_id, dtype: int64


Дополнительная обработка значений в данном столбце не требуется. Значения соответствуют столбцу `education`. В дальнейшем будет составлен словарь, а столбец `education` будет удален. 

In [166]:
# Выводим все уникальные значения по столбцу `family_status`

print(df['family_status'].value_counts())

женат / замужем          12351
гражданский брак          4172
Не женат / не замужем     2808
в разводе                 1191
вдовец / вдова             956
Name: family_status, dtype: int64


In [167]:
# Приведем значения столбца `family_status` к нижнему регистру

df['family_status'] = df['family_status'].str.lower()
display(df['family_status'].value_counts()) 

женат / замужем          12351
гражданский брак          4172
не женат / не замужем     2808
в разводе                 1191
вдовец / вдова             956
Name: family_status, dtype: int64

In [168]:
# Выводим все уникальные значения по столбцу `family_status_id`

print(df['family_status_id'].value_counts())

0    12351
1     4172
4     2808
3     1191
2      956
Name: family_status_id, dtype: int64


Дополнительная обработка значений в данном столбце не требуется. Значения соответствуют столбцу `family_status`. В дальнейшем будет составлен словарь, а столбец `family_status` будет удален. 

In [169]:
# Выводим все уникальные значения по столбцу `gender`

print(df['gender'].value_counts())

F      14201
M       7276
XNA        1
Name: gender, dtype: int64


Мы видим, что одно из значений в колонке имеет значение "XNA". Сложно точно сказать, что именно стоит за этим значением. Можно предположить, что пользователь либо отказался вводить пол, либо это какое-то обозначение одного из гендеров, к которым он себя причислил. Однако имеет место быть и технологической ошибке. Выведем всю строку для проверки корректности остальных данных.

In [170]:
# Выводим строчку со значением XNA в столбце `gender`

display(df.loc[df.loc[:,'gender'] == 'XNA'])

children  days_employed  dob_years            education  education_id  \
10675         0          -2358         24  неоконченное высшее             2   

          family_status  family_status_id gender income_type   debt  \
10675  гражданский брак                 1    XNA   компаньон  False   

       total_income               purpose  
10675        203905  покупка недвижимости

Остальные значения в строке не мешают нашим исследованиям. Заменять пол XNA на М или Ж не имеет смысла, т.к. это не повлияет на наши исследования. Оставляем столбец без изменений.

In [171]:
# Выводим все уникальные значения по столбцу `income_type`

print(df['income_type'].value_counts())

сотрудник          11093
компаньон           5076
пенсионер           3848
госслужащий         1455
безработный            2
предприниматель        2
студент                1
в декрете              1
Name: income_type, dtype: int64


Данные в столбце `income_type` являются корректными и не требует дополнительной обработки. Оставляет столбец без изменений.

In [172]:
# Выводим все уникальные значения по столбцу `debt`

print(df['debt'].value_counts())

False    19738
True      1740
Name: debt, dtype: int64


Данные в столбце `debt` являются корректными и не требует дополнительной обработки. Оставляет столбец без изменений.

* 0 - задолженностей по возврату кредитов не имел
* 1 - задолженность по возрату кредитов есть в кредитной истории

В случае со столбцом `total_income` воспользоваться методами .value_count() или .unique() не представляется возможным ввиду большого пула различных данных. Для поиска аномалий определим минимальный и максимальный уровень дохода.

In [173]:
# Находим минимальный и максимальный уровень дохода заемщиков

print('Максимальный уровень дохода заемщика: ', df['total_income'].max())
print('Минимальный уровень дохода заемщика: ', df['total_income'].min())

Максимальный уровень дохода заемщика:  2265604
Минимальный уровень дохода заемщика:  20667


По полученным данным можем сделать вывод, что отсутствуют отрицательные или близкие к бесконечности значения, что косвенно сведетельсвует от корректности данных в столбце `total_income`

In [174]:
# Выводим все уникальные значения по столбцу `purpose`

print(df['purpose'].value_counts())

свадьба                                   797
на проведение свадьбы                     776
сыграть свадьбу                           773
операции с недвижимостью                  674
покупка коммерческой недвижимости         664
операции с жильем                         652
покупка жилья для сдачи                   652
операции с коммерческой недвижимостью     650
жилье                                     646
покупка жилья                             644
покупка жилья для семьи                   641
строительство собственной недвижимости    634
недвижимость                              633
операции со своей недвижимостью           627
строительство жилой недвижимости          623
покупка недвижимости                      621
строительство недвижимости                620
покупка своего жилья                      619
ремонт жилью                              611
покупка жилой недвижимости                605
на покупку своего автомобиля              504
заняться высшим образованием      

Из полученных данных видим большое число неявных дубликатов, которые требуется устранить. 

Устранять эти дубликаты будем путем создания нового столбца `purpose_category`, который будет иметь всего 4 категории:
* операции с автомобилем
* операции с недвижимостью
* проведение свадьбы
* получение образования

Проделаем эту операцию на одном из следующих шагов.


In [175]:
# Посчитаем явных количество дубликатов в таблице

print('Дубликатов в таблице:', df.duplicated().sum())

Дубликатов в таблице: 71


In [176]:
# Удаляем все явные дубликаты в таблице

df = df.drop_duplicates().reset_index(drop=True)

In [177]:
# Проверяем число явных дубликатов после удаления

print('Дубликатов в таблице после обработки:', df.duplicated().sum())

Дубликатов в таблице после обработки: 0


<div style="border:solid green 2px; padding: 20px"> 
    
**Комментарий ревьюера**

**#ок**

Отлично, все ошибки исправлены. Данные теперь достаточно чистые для дальнейшей работы

</div>

<div style="border:solid green 2px; padding: 20px"> 
    
**Комментарий ревьюера - 2**

Отлично, теперь все верно
    
</div>

**Выводы**

В результате предыдущих шагов была проделана большая часть работы по предъобработке данных, а именно:
* Были удалены и/или исправлены явные и неявные дубликаты
* Были изменены типы данных в некоторых столбцах
* Были найдены аномалии и приняты решения об их дальнейшей обработке
* Были назначены планы на дальнейшее исследование

# Шаг 2.5. Формирование дополнительных датафреймов словарей, декомпозиция исходного датафрейма.

Создадим два новых датафрема, содержащих в себе:
* education_id и education — в первом;
* family_status_id и family_status — во втором.

In [178]:
# Создадим два новых датафрейма "словаря"

df_education = df[['education_id', 'education']]
df_family = df[['family_status_id', 'family_status']]

df_education = df_education.drop_duplicates().reset_index(drop=True)
df_family = df_family.drop_duplicates().reset_index(drop=True)

display(df_education.head(5))
display(df_family.head(5))

education_id            education
0             0               высшее
1             1              среднее
2             2  неоконченное высшее
3             3            начальное
4             4       ученая степень

family_status_id          family_status
0                 0        женат / замужем
1                 1       гражданский брак
2                 2         вдовец / вдова
3                 3              в разводе
4                 4  не женат / не замужем

Мы получили два датафрейма-"словаря", к которым мы сможем обращаться по идентификатору в дальнейшем.

Удалим из исходного датафрейма столбцы `education_id` и `family_status_id`.

In [179]:
# Удаляем столбцы `education_id` и `family_status_id`. Проверяем результат.

df = df.drop(['education_id', 'family_status_id'], axis = 1)
df.head(5)

children  days_employed  dob_years education     family_status gender  \
0         1          -8437         42    высшее   женат / замужем      F   
1         1          -4024         36   среднее   женат / замужем      F   
2         0          -5623         33   среднее   женат / замужем      M   
3         3          -4124         32   среднее   женат / замужем      M   
4         0         340266         53   среднее  гражданский брак      F   

  income_type   debt  total_income                     purpose  
0   сотрудник  False        253875               покупка жилья  
1   сотрудник  False        112080     приобретение автомобиля  
2   сотрудник  False        145885               покупка жилья  
3   сотрудник  False        267628  дополнительное образование  
4   пенсионер  False        158616             сыграть свадьбу

**Вывод**

Мы избавились от двух столбов исходного датафрейма и создали для них два дополнительных датафрейма. Таблица приобрела более читаемый вид.

<div style="border:solid green 2px; padding: 20px"> 
    
**Комментарий ревьюера**

**#ок**

Отлично, словари правильно выделены
</div>

# Шаг 2.6. Категоризация дохода.

Необходимо категоризировать столбец `total_income` на 5 категорий:
* 0–30000 — 'E';
* 30001–50000 — 'D';
* 50001–200000 — 'C';
* 200001–1000000 — 'B';
* 1000001 и выше — 'A'.

Для этого напишем функцию одной строки, которая создаст новый столбец `total_income_category` и распределит клиентские доходы на обозначенные категории.

In [180]:
def income_group(income):
    try:
        if income <= 30000:
            return 'E'
        if 30001 <= income <= 50000:
            return 'D'
        if 50001 <= income <= 200000:
            return 'C'
        if 200001 <= income <= 1000000:
            return 'B'
        if income >= 1000001:
            return 'A'
        else:
            return 'unknown category'
    except:
        return 'Неизвестная ошибка. Проверьте функцию.'
    
df['total_income_category'] = df['total_income'].apply(income_group)

In [181]:
# Проверяем корректность работу функции

print(df['total_income_category'].value_counts())
df.head(5)

C    15978
B     5033
D      349
A       25
E       22
Name: total_income_category, dtype: int64


children  days_employed  dob_years education     family_status gender  \
0         1          -8437         42    высшее   женат / замужем      F   
1         1          -4024         36   среднее   женат / замужем      F   
2         0          -5623         33   среднее   женат / замужем      M   
3         3          -4124         32   среднее   женат / замужем      M   
4         0         340266         53   среднее  гражданский брак      F   

  income_type   debt  total_income                     purpose  \
0   сотрудник  False        253875               покупка жилья   
1   сотрудник  False        112080     приобретение автомобиля   
2   сотрудник  False        145885               покупка жилья   
3   сотрудник  False        267628  дополнительное образование   
4   пенсионер  False        158616             сыграть свадьбу   

  total_income_category  
0                     B  
1                     C  
2                     C  
3                     B  
4                     C

**Вывод**

Функция работает корректно. Ошибок в работе функции не выявлено, т.к. в случае нераспознавания категории дохода появилась бы дополнительная категория `unknown category`, а в случае ошибки выполнения функции сработала бы ветвь except и была бы выведена соответствующая ошибка.

# Шаг 2.7. Категоризация целей кредита.

Необходимо категоризировать столбец `purpose` на 4 категории:
* операции с автомобилем
* операции с недвижимостью
* проведение свадьбы
* получение образования

Для этого напишем функцию одной строки, которая создаст новый столбец `purpose_category` и распределит клиентские кредитные цели на обозначенные категории путем использования ключевых слов.

In [182]:
# Создаем функцию одной строки, которая категорезует все причины займов клиентов на 4 категории

def purpose_group(purpose):
    try:
        if 'свадьб' in purpose:
            return 'проведение свадьбы'
        if 'авто' in purpose:
            return 'операции с автомобилем'
        if 'образов' in purpose:
            return 'получение образования'
        if ('недвиж' in purpose) or ('жиль' in purpose):
            return 'операции с недвижимостью'
        else:
            return 'unknown category'
    except:
        return 'Неизвестная ошибка. Проверьте функцию.'
    
df['purpose_category'] = df['purpose'].apply(purpose_group)

In [183]:
# Проверяем корректность работу функции

print(df['purpose_category'].value_counts())
df.head(5)

операции с недвижимостью    10787
операции с автомобилем       4295
получение образования        4003
проведение свадьбы           2322
Name: purpose_category, dtype: int64


children  days_employed  dob_years education     family_status gender  \
0         1          -8437         42    высшее   женат / замужем      F   
1         1          -4024         36   среднее   женат / замужем      F   
2         0          -5623         33   среднее   женат / замужем      M   
3         3          -4124         32   среднее   женат / замужем      M   
4         0         340266         53   среднее  гражданский брак      F   

  income_type   debt  total_income                     purpose  \
0   сотрудник  False        253875               покупка жилья   
1   сотрудник  False        112080     приобретение автомобиля   
2   сотрудник  False        145885               покупка жилья   
3   сотрудник  False        267628  дополнительное образование   
4   пенсионер  False        158616             сыграть свадьбу   

  total_income_category          purpose_category  
0                     B  операции с недвижимостью  
1                     C    операции с автомобилем  
2                     C  операции с недвижимостью  
3                     B     получение образования  
4                     C        проведение свадьбы

**Вывод**

Функция работает корректно. Ошибок в работе функции не выявлено, т.к. в случае нераспознавания причины займа появилась бы дополнительная категория `unknown category`, а в случае ошибки выполнения функции сработала бы ветвь except и была бы выведена соответствующая ошибка.

<div style="border:solid green 2px; padding: 20px"> 
    
**Комментарий ревьюера**

**#ок**

Отлично, требуемые по заданию категории выделены правильно, молодец
</div>

<div style="border:solid gold 2px; padding: 20px"> 
    
**Комментарий ревьюера**

**#можно лучше**

также стоило выделить категории по количеству детей (хотя это напрямую в задании не сказано). Если ты посмотришь далее, то при оценки вероятности невозврата у на появились категории с 9 и 41 пользователем. Вообще говоря такие группы анализировать статистически не совсем верно. правильнее было бы укрупнить группы, чтобы в каждой было хотя бы 100 пользователей
</div>

<div class="alert alert-info">
<b> Комментарий от студента - 2</b>
    
Идею понял, а каким образом в данном случае стоило бы укрупнить группы, чтобы не повлиять на задачи исследования? Допустим, объедени я людей с 4, 5 и 20 детьми в одну группу "4 и более" - насколько это было бы корректно?
</div>

<div style="border:solid green 2px; padding: 20px"> 
    
**Комментарий ревьюера - 2**

В данном случае можно сделать просто категории  сдетьми/без детей. Вообще будет корректно любое разделение, где группы получатся больше 100. Но тут явный перекос в сторону 0 детей и только если все остальные объединить, то получим равнозначные группы
    
</div>

# Ответы на вопросы.

Данные прошли все необходимые этапы предобработки и готовы к исследованиям. Нам необходимо ответить на несколько вопросов:
1. Есть ли зависимость между количеством детей и возвратом кредита в срок?
2. Есть ли зависимость между семейным положением и возвратом кредита в срок?
3. Есть ли зависимость между уровнем дохода и возвратом кредита в срок?
4. Как разные цели кредита влияют на его возврат в срок?

Поочередно проанализируем данные, чтобы получить ответы на эти вопросы.

**Вопрос 1: Есть ли зависимость между количеством детей и возвратом кредита в срок?**

Для ответа на данный вопрос напишем функцию, которая опираясь на наши данные вернет процент добросовестных вкладчиков с N количеством детей, относительно общего числа вкладчиков с числом детей N.

In [184]:
# Функция для расчета процента добросовестных заемщиков, относительно общего числа кредиторов с заданным числом детей
# Функция получает на вход количество детей и debt (0 - просрочек не было, 1 - есть просрочки)
# Функция находит строчки с заданным числом детей, считает их количество.
# Затем функция находит число детей в категории debt = 0 и считает их количество.
# Далее происходит округление и расчет отношения числа добросовестных заёмщиков к общему числу с выбранным числом детей.
# И далее с помощью цикла for функция вызывается нужное количество раз.

def children_count(child_count, debt_number):
    child_list = df.loc[df.loc[:,'children'] == child_count]
    child_list_count_total = child_list.shape[0]
    child_list = child_list.loc[child_list.loc[:,'debt'] == debt_number]
    child_list_count = child_list.shape[0]
    child_relationship = round((child_list_count / child_list_count_total * 100), 1)
    child_relationship = 'Процент добросовестных заёмщиком с ' + str(child_count) + ' детей(детьми): ' + str(child_relationship) + '%'
    return child_relationship

children_list = [0, 1, 2, 3, 4, 5, 20]

for children in children_list:
    print(children_count(children, 0))

Процент добросовестных заёмщиком с 0 детей(детьми): 92.5%
Процент добросовестных заёмщиком с 1 детей(детьми): 90.8%
Процент добросовестных заёмщиком с 2 детей(детьми): 90.5%
Процент добросовестных заёмщиком с 3 детей(детьми): 91.8%
Процент добросовестных заёмщиком с 4 детей(детьми): 90.2%
Процент добросовестных заёмщиком с 5 детей(детьми): 100.0%
Процент добросовестных заёмщиком с 20 детей(детьми): 89.5%


In [185]:
# Посчитаем число строк в таблице, где количество детей у заёмщика - 5

child_list = df.loc[df.loc[:,'children'] == 5]
child_list_count_total = child_list.shape[0]
print('Общее число строк с базе с пятью детьми:', child_list_count_total)

Общее число строк с базе с пятью детьми: 9


In [207]:
# Строим сводную таблицу с вероятностью невозврата кредита в срок

df_income = df.pivot_table(index=['children'], values='debt', aggfunc='mean').sort_values(by='debt')
df_income['debt'] = round(df_income['debt'] * 100, 1)
df_income = df_income.rename(
            columns={
                'debt': 'Вероятность невозврата в срок (%)',
            }
)


display(df_income)

Вероятность невозврата в срок (%)
children                                   
5                                       0.0
0                                       7.5
3                                       8.2
1                                       9.2
2                                       9.5
4                                       9.8
20                                     10.5

**Вывод 1:**

Отвечая на вопрос "Есть ли зависимость между количеством детей и возвратом кредита в срок?" можно сделать следующие выводы:
* Люди без детей возвращают кредиты в срок с наибольшей долей вероятности (7.5% невозвратов)
* Люди с 1 и 2 детьми возвращают кредиты в срок приблизительно с одинаковой вероятностью (9.2 и 9.5 процентов невозвратов соответственно)
* Люди с 4 и 20 детьми возвращают кредиты в срок приблизительно с одинаковой вероятностью (9.8 и 10.5 процентов невозвратов соответственно)
* Люди с 3 детьми возвращают кредит с долей вероятности выше, чем с 1 и 2, но ниже, чем люди без детей (8.2% невозвратов)
* Люди с 5 детьми возвращают кредит со 100% долей вероятности, однако стоит отметить, что результаты для данного количества детей нельзя справедливо оценить, т.к. в базе присутствует слишком мало данных по таким заёмщикам. Общее число таких строк - 9.


**Вопрос 2: Есть ли зависимость между семейным положением и возвратом кредита в срок?**

In [186]:
# Аналогичная функция для расчета процента добросовестных заемщиков по семейному положению

def family_count(family_meaning, debt_number):
    family_list = df.loc[df.loc[:,'family_status'] == family_meaning]
    family_list_count_total = family_list.shape[0]
    family_list = family_list.loc[family_list.loc[:,'debt'] == debt_number]
    family_list_count = family_list.shape[0]
    family_relationship = round((family_list_count / family_list_count_total * 100), 1)
    family_relationship = 'Процент добросовестных заёмщиком в семейном положении ' + str(family_meaning) + ': ' + str(family_relationship) + '%'
    return family_relationship

family_list = ['женат / замужем', 'гражданский брак', 'не женат / не замужем', 'в разводе', 'вдовец / вдова']

for family in family_list:
    print(family_count(family, 0))

Процент добросовестных заёмщиком в семейном положении женат / замужем: 92.4%
Процент добросовестных заёмщиком в семейном положении гражданский брак: 90.6%
Процент добросовестных заёмщиком в семейном положении не женат / не замужем: 90.2%
Процент добросовестных заёмщиком в семейном положении в разводе: 92.9%
Процент добросовестных заёмщиком в семейном положении вдовец / вдова: 93.4%


In [206]:
# Строим сводную таблицу с вероятностью невозврата кредита в срок

df_income = df.pivot_table(index=['family_status'], values='debt', aggfunc='mean').sort_values(by='debt')
df_income['debt'] = round(df_income['debt'] * 100, 1)
df_income = df_income.rename(
            columns={
                'debt': 'Вероятность невозврата в срок (%)',
            }
)


display(df_income)

Вероятность невозврата в срок (%)
family_status                                           
вдовец / вдова                                       6.6
в разводе                                            7.1
женат / замужем                                      7.6
гражданский брак                                     9.4
не женат / не замужем                                9.8

**Вывод 2:**

Отвечая на вопрос "Есть ли зависимость между семейным положением и возвратом кредита в срок?" можно сделать следующие выводы:

* Наиболее добросовестными заемщиками являются люди в семейном положении "вдовец/вдова" - 6.6%
* Заёмщики с семейным положением "в разводе" - 7.1%
* Заёмщики с семейным положением "женат/замужем" - 7.6%
* Заёмщики с семейным положением "гражданский брак" - 9.4%
* Заёмщики с семейным положнием "не женат/не замужем" - 9.8%

Таким образом можно сделать вывод, что наиболее добросовестными заёмщиками являются люди с семейным положением "вдовец/вдова" и "в разводе", а категория людей с семейным положением "гражданский брак" и "не женат/не замужем" имеют наиболее низкие показатели.

**Вопрос 3: Есть ли зависимость между уровнем дохода и возвратом кредита в срок?**

In [187]:
# Аналогичная функция для расчета процента добросовестных заемщиков по уровню дохода

def income_count(income_meaning, debt_number):
    income_list = df.loc[df.loc[:,'total_income_category'] == income_meaning]
    income_list_count_total = income_list.shape[0]
    income_list = income_list.loc[income_list.loc[:,'debt'] == debt_number]
    income_list_count = income_list.shape[0]
    income_relationship = round((income_list_count / income_list_count_total * 100), 1)
    income_relationship = 'Процент добросовестных заёмщиков с уровнем дохода в категории ' + str(income_meaning) + ': ' + str(income_relationship) + '%'
    return income_relationship

income_list = ['A', 'B', 'C', 'D', 'E']

for income in income_list:
    print(income_count(income, 0))

Процент добросовестных заёмщиков с уровнем дохода в категории A: 92.0%
Процент добросовестных заёмщиков с уровнем дохода в категории B: 92.9%
Процент добросовестных заёмщиков с уровнем дохода в категории C: 91.5%
Процент добросовестных заёмщиков с уровнем дохода в категории D: 94.0%
Процент добросовестных заёмщиков с уровнем дохода в категории E: 90.9%


In [204]:
# Строим сводную таблицу с вероятностью невозврата кредита в срок

df_income = df.pivot_table(index=['total_income_category'], values='debt', aggfunc='mean').sort_values(by='debt')
df_income['debt'] = round(df_income['debt'] * 100, 1)
df_income = df_income.rename(
            columns={
                'debt': 'Вероятность невозврата в срок (%)',
            }
)


display(df_income)

Вероятность невозврата в срок (%)
total_income_category                                   
D                                                    6.0
B                                                    7.1
A                                                    8.0
C                                                    8.5
E                                                    9.1

**Вывод 3:**

Отвечая на вопрос "Есть ли зависимость между уровнем дохода и возвратом кредита в срок?" можно сделать следующие выводы:

* Люди с категорией дохода D (30001 - 50000) имеют самый низкий показатель невозврата в срок - 6.0%
* Люди с категорией дохода E (0 - 30000) имеют самый высокий показатель невозврата в срок - 9.1%
* Люди с категориями дохода C, A, B имеют средние уровни показателя невозврата в срок - 8.5, 8.0 и 7.1 процентов соответственно

Таким образом можно сделать вывод, что заёмщики с категорией дохода D имеют самый низкий показатель невозврата, а люди с категорией E - наиболее высокий.

**Вопрос 4: Как разные цели кредита влияют на его возврат в срок?**

In [188]:
# Аналогичная функция для расчета процента добросовестных заемщиков по цели кредита

def purpose_count(purpose_meaning, debt_number):
    purpose_list = df.loc[df.loc[:,'purpose_category'] == purpose_meaning]
    purpose_list_count_total = purpose_list.shape[0]
    purpose_list = purpose_list.loc[purpose_list.loc[:,'debt'] == debt_number]
    purpose_list_count = purpose_list.shape[0]
    purpose_relationship = round((purpose_list_count / purpose_list_count_total * 100), 1)
    purpose_relationship = 'Процент добросовестных заёмщиков с категорией кредита ' + str(purpose_meaning) + ': ' + str(purpose_relationship) + '%'
    return purpose_relationship

purpose_list = ['операции с недвижимостью', 'операции с автомобилем', 'получение образования', 'проведение свадьбы']

for purpose in purpose_list:
    print(purpose_count(purpose, 0))

Процент добросовестных заёмщиков с категорией кредита операции с недвижимостью: 92.8%
Процент добросовестных заёмщиков с категорией кредита операции с автомобилем: 90.6%
Процент добросовестных заёмщиков с категорией кредита получение образования: 90.8%
Процент добросовестных заёмщиков с категорией кредита проведение свадьбы: 92.0%


In [203]:
# Строим сводную таблицу с вероятностью невозврата кредита в срок

df_income = df.pivot_table(index=['purpose_category'], values='debt', aggfunc='mean').sort_values(by='debt')
df_income['debt'] = round(df_income['debt'] * 100, 1)
df_income = df_income.rename(
            columns={
                'debt': 'Вероятность невозврата в срок (%)',
            }
)


display(df_income)

Вероятность невозврата в срок (%)
purpose_category                                           
операции с недвижимостью                                7.2
проведение свадьбы                                      8.0
получение образования                                   9.2
операции с автомобилем                                  9.4

**Вывод 4:**

Отвечая на вопрос "Как разные цели кредита влияют на его возврат в срок?" можно сделать следующие выводы:

* Заёмщики с потребностью кредита на операции с недвижимостью имеют наименьший показатель невозврата - 7.2%
* Заёмщики с потребностью кредита на проведение свадьбы имеют средний показатель невозврата - 8.0%
* Заёмщики с потребностью кредита на операции с автомобилем и получением образования имеют наиболее высокие показатели невозврата - 9.4 и 9.2 процентов соответственно.

Таким образом, люди с целью кредита на операции с недвижимостью имеют наименьший показатель невозврата, а люди, берующие кредит на операции с автомобилем и получением образования наибольший показатель невозврата.

<div style="border:solid red 2px; padding: 20px"> 
    
**Комментарий ревьюера**

**#ошибка**

Ты в целом считаешь верные проценты (хотя у банкиров больше принято оценивать вероятность невозврата - риск), но не самым оптимальным методом.
    
Получи, пожалуйста те же проценты только с помощью pivot_table. Не правильно для подобных задач использовать циклы, все должно быть выведено в таблицах. Соответственно для каждого вопроса нужно получить таблицу с двумя основными колонками (если надо можно дополнительные колонки вывести): категории и вероятность возврата/просрочки.
    
Также нужно сделать сортировку полученных значений и отформатировать их в виде процентов для лучшей считываемости
</div>

<div class="alert alert-info">
<b> Комментарий от студента - 2</b>
    
Замечание понял, исправил для каждого из пунктов. Вариант расчета с помощью функции не удалял, чтобы легче было вспомнить прошлую итерацию. Возникло несколько вопросов:
1. Как добавить ко всем значениям столбца 'debt' знак %. Пытался преобразовать столбец в строку, а затем приклеить к нему знак процента. Но мне кажется, что это какой-то мамонтовский метод. Подскажи, может есть более лакончиный способ это сделать? И стоит ли вообще единицы измерения указывать непосредственно в значениях ячеек?
2. Можно ли как-то переименовать первый столбец в получившейся таблице? Если к тому варианту, что я использовал для переименования столбца 'debt' добавить индексируемый столбец, то он не меняется. Может есть какая-то особенность о которой я не знаю?
    

Если критических замечаний не будет, то буду рад получить ответы на эти вопросы в желтой ячейке для запоминания и применения в будущем.
</div>

<div style="border:solid green 2px; padding: 20px"> 
    
**Комментарий ревьюера - 2**

Теперь все верно

1. Можно оставить как есть, подписть % в столбце вполне можно, но можно использовать `styling` https://pandas.pydata.org/pandas-docs/version/1.1/user_guide/style.html
    

2. Можно будет менять столбцы, если сделать reset_index. Сейчас первый столбец - это индекс
    
    df_income = df.pivot_table(index=['purpose_category'], values='debt', aggfunc='mean').sort_values(by='debt').reset_index()
    
Далее с таблицей можно работать как с простой линейной
</div>

# Общий вывод:

В результате данной работы мы получили ответы на вопросы заказчика:
* Есть ли зависимость между количеством детей и возвратом кредита в срок?
* Есть ли зависимость между семейным положением и возвратом кредита в срок?
* Есть ли зависимость между уровнем дохода и возвратом кредита в срок?
* Как разные цели кредита влияют на его возврат в срок?

Люди с разным количеством детей, различным семейным положением, уровенем дохода и целью кредита по-разному влияют на критерий "возврат в срок".
* Люди с целью кредита на операции с недвижимостью имеют наибольший показатель возврата, а люди, берующие кредит на операции с автомобилем и получением образования наименьший показатель возврата.
* Заёмщики с категорией дохода D имеют самый высокий показатель возврата, а люди с категорией E - наиболее низкий.
* Наиболее добросовестными заёмщиками являются люди с семейным положением "вдовец/вдова" и "в разводе", а категория людей с семейным положением "гражданский брак" и "не женат/не замужем" имеют наиболее низкие показатели.
* Люди без детей возвращают кредиты в срок с наибольшей долей вероятности, а люди с 4 и 20 детьми возвращают кредиты в срок с наиболее низкой долей вероятности.

<div class="alert alert-info">
<b> Комментарий от студента </b>
    
Большое спасибо, что потратили время на мою работу. Буду рад Вашим пожеланиям и замечаниям. Хорошего дня/вечера или даже ночи :)
</div>

<div style="border:solid green 2px; padding: 20px"> 

**Общий комментарий по проекту**

Ты отлично поработал над проектом, молодец. Ты сделал хороший самостоятельный анализ
    
Ты молодец, отлично справляешься с программой. 
    
Поправь, пожалуйста небольшие неточности, чтобы я могла принять проект.

Удачи!
</div>

<div class="alert alert-info">
<b> Комментарий от студента - 2 </b>
    
Большое спасибо за проверку проекта! Хорошего вечера/дня/ночи/утра :)
</div>